### PR TITLE
Solver deep

### PR DESCRIPTION
**Pull request by chris-mrn**
we can now have data augmentation with each of our solvers

### CODE REVIEW

**Review by bruAristimunha — APPROVED** [review]

**Comment by bruAristimunha** [comment]
Can we merge this iteration @chris-mrn and @tomMoral?

File: README.rst
Comment (tomMoral) [inline README.rst:36]: Not sure we want to recommend `DEBUG` mode? And usually, we consider that the user just cloned the benchmark so they are in the parent directory.
```suggestion
    $ benchopt run benchmark_bci -s MDM -d BNCI -r 1 -n 1
```

File: benchmark_utils/__init__.py
Comment (tomMoral) [inline benchmark_utils/__init__.py:42]: simpler
```suggestion
    for Xi in X:
        X0.append(Xi)
```

File: benchmark_utils/__init__.py
Comment (tomMoral) [inline benchmark_utils/__init__.py:71]: This seems very hard coded, and it will break if we extend to other paradigm.
Can't you get this by passing the paradigm as an argument of the function and creating the mapping on the fly?

File: benchmark_utils/__init__.py
Comment (tomMoral) [inline benchmark_utils/__init__.py:113]: Shouldn't this function be in `transform.py?`

File: benchmark_utils/__init__.py
Comment (tomMoral) [inline benchmark_utils/__init__.py:130]: ditto

File: benchmark_utils/__init__.py
Comment (tomMoral) [reply benchmark_utils/__init__.py:113]: it is actually duplicated.

File: objective.py
Comment (tomMoral) [inline objective.py:28]: Don't you need skorch too?

File: objective.py
Comment (tomMoral) [inline objective.py:96]: :rofl: you killed me here
Avoid french in the code, and in particular accents
```suggestion
            for key in data_split_session.items():
                if key[0] != str(session_test):
                    data_session_train += data_split_session[key[0]]
```

**Review by tomMoral — COMMENTED** [review]
A first set of comments

File: objective.py
Comment (bruAristimunha) [reply objective.py]: Not need, because braindecode already install skorch.

**Comment by bruAristimunha** [comment]
Waiting for your next review @tomMoral and @matthieutrs ;)

File: benchmark_utils/__init__.py
Comment (tomMoral) [inline benchmark_utils/__init__.py:11]: I don't see why you need the import context?

File: benchmark_utils/__init__.py
Comment (tomMoral) [inline benchmark_utils/__init__.py:12]: avoid using `noqa` when exposing function in a module.
good practice is to put the exposed function in a `__all__` variable:

`__all__ = ['smooth_time_mask', ...]`

and this works with the linter.

File: benchmark_utils/dataset.py
Comment (tomMoral) [inline benchmark_utils/dataset.py:40]: isn't there this mapping info directly in the paradigm? this feels weird to hard code this in the function as this week mean we need to maintain a list for every paradigm we want to use

Comment (tomMoral) [reply benchmark_utils/dataset.py:40]: at least if we have a list with the babes of the events, we could generate this on the fly

File: benchmark_utils/dataset.py
Comment (tomMoral) [inline benchmark_utils/dataset.py:55]: this should be defined by each dataset no?

File: benchmark_utils/transformation.py
Comment (tomMoral) [inline benchmark_utils/transformation.py:99]: avoid having many unrelated changes in a single PR, this makes it hard to review.

File: objective.py
Comment (tomMoral) [inline objective.py:28]: why put this line this age but just directly the string?

File: objective.py
Comment (tomMoral) [inline objective.py:34]: align this closing bracket with the beginning of the line where the matching `[` is.

File: objective.py
Comment (tomMoral) [inline objective.py:42]: why do you need `sfreq`? can't you recover it from the dataset?

File: objective.py
Comment (tomMoral) [inline objective.py:47]:
```suggestion
```

File: objective.py
Comment (tomMoral) [inline objective.py:121]: as we continue several metrics here, we should move to computing the test prediction once and then compute all the metrics like BAS.

File: objective.py
Comment (tomMoral) [inline objective.py:145]: don't put this feature in this pr and avoid pushing note comments on GitHub

File: pipelines/WTRCSP.py
Comment (tomMoral) [inline pipelines/WTRCSP.py:1]: can the solver take both yml and py files?

File: solvers/CSPlda.py
Comment (tomMoral) [inline solvers/CSPlda.py:28]: why the parenthesis?

File: solvers/CSPlda.py
Comment (tomMoral) [inline solvers/CSPlda.py:33]: isn't torch a dependency of `braindecode`?

Comment (tomMoral) [reply solvers/CSPlda.py:33]: also, it feels weird to need `pytorch ` for the baseline model. wouldn't it be possible to avoid this?

File: solvers/Dummyclassifier.py
Comment (tomMoral) [inline solvers/Dummyclassifier.py:1]: why is this for useful?

File: solvers/KN.py
Comment (tomMoral) [inline solvers/KN.py:60]: &
There is a lot of duplicated code.
Could you do a base augmented solver class with the mechanism to augment in `benchmark_utils/augmented_dataset.py` and make the other solvers depend on it?

File: benchmark_utils/__init__.py
Comment (bruAristimunha) [reply benchmark_utils/__init__.py:12]: Nice, I didn't know this. Thank you!

File: solvers/KN.py
Comment (bruAristimunha) [reply solvers/KN.py:60]: Yes

File: solvers/CSPlda.py
Comment (bruAristimunha) [reply solvers/CSPlda.py]: Yes! Removed

File: pipelines/WTRCSP.py
Comment (bruAristimunha) [reply pipelines/WTRCSP.py:1]: Yes =)

Comment (bruAristimunha) [reply pipelines/WTRCSP.py:1]: Good idea!

File: solvers/Dummyclassifier.py
Comment (bruAristimunha) [reply solvers/Dummyclassifier.py]: Moved Dummy Classifier to Moabb Pipelines

File: benchmark_utils/transformation.py
Comment (bruAristimunha) [reply benchmark_utils/transformation.py]: okay, noted!

Comment (bruAristimunha) [reply benchmark_utils/transformation.py]: We will improve this new augmentation method in another PR, as soon as this one is merged

File: objective.py
Comment (bruAristimunha) [reply objective.py:117]: Sorry, this was the only point that I didn't catch

File: benchmark_utils/dataset.py
Comment (bruAristimunha) [reply benchmark_utils/dataset.py:94]: This is more of a convention of what we will adopt as left and right order; we can move that somewhere better. We will have to adapt which classes we will use for each dataset and paradigm. The number of classes is always variable per dataset, and no information in the dataset will guarantee alignment and/or order.

I need some help with this point to improve this design.

**Comment by bruAristimunha** [comment]
Hi @tomMoral, @matthieutrs and @chris-mrn!

Thank you so much for your last review. I loved! I think I implemented almost everything that was asked, with the exception of two points, which I left in the open. I was wondering, could we merge the PR a bit so we can focus on cross-validation?

File: benchmark_utils/dataset.py
Comment (matthieutrs) [reply benchmark_utils/dataset.py:94]: In MOABB this is available in [here](https://github.com/NeuroTechX/moabb/blob/develop/moabb/datasets/bnci.py#L797), I guess similarly it can be accessed through the [mapping](https://github.com/braindecode/braindecode/blob/master/braindecode/preprocessing/windowers.py#L119) in braindecode/mne as well... Maybe it could be use these?

File: objective.py
Comment (matthieutrs) [reply objective.py:44]: Not sure I understand this point: isn't `dataset` a braindecode object?

File: pipelines/CSP.yml
Comment (chris-mrn) [reply pipelines/CSP.yml:1]: if it is the case, why did you get rid of CSPLDA?

**Review by chris-mrn — COMMENTED** [review]
review bruno

File: pipelines/CSP.yml
Comment (bruAristimunha) [reply pipelines/CSP.yml:1]: Hi!
This implementation is different from what would be a "normal CSP". It is applied in the covariance matrix, as per [Pyriemann's](https://pyriemann.readthedocs.io/en/latest/generated/pyriemann.spatialfilters.CSP.html#pyriemann.spatialfilters.CSP) documentation.

As we have a CSP implementation directly in the signal, I preferred the rename to clarify that we have two CSPs, but they are applied to different data types.

File: solvers/CSPlda.py
Comment (bruAristimunha) [reply solvers/CSPlda.py:1]: Here we are directly applying CSP, and mne asks for this numerical conversion

**Review by bruAristimunha — APPROVED** [review]

File: datasets/Zhou.py
Comment (bruAristimunha) [reply datasets/Zhou.py:1]: No hard reason, but I think we can do something smarter about the dataset and paradigms. Possibly in another PR.

**Comment by bruAristimunha** [comment]
I think I've addressed all comments and reviews. The items that were not resolved, I think we can address them in other PR. Can we merge @tomMoral, @matthieutrs and @chris-mrn?

**Comment by matthieutrs** [comment]
Yes, I think we can merge :)

**Comment by tomMoral** [comment]
Ad this was beginning to be blocking, I merged as is. 

But could you open issues for the un resolved comments that need to be addressed (one at a time this time 😅).

**Pull request merged**
